### PR TITLE
usersession/client: add a client library for the user session agent

### PR DESCRIPTION
**Pull request by jhenstridge**
This package provides a client library for the user session agent.  It is roughly modelled after the snapd client library `client`, with a few major differences:

1. There can be multiple session agents available on the system, so method calls fan out to all agents and collect the results.

2. All methods take a `context.Context` parameter, making it possible to set a deadline or otherwise cancel the request.  This is important, since the common use case will be code running as root making calls to an agent controlled by a regular user.

### CODE REVIEW

File: usersession/client/client.go [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+	var resultErr Error
+	err := json.Unmarshal(resp.Result, &resultErr)
+	if err != nil || resultErr.Message == "" {
+		resp.err = fmt.Errorf("server error: %q", http.StatusText(resp.statusCode))
+	} else {
+		resp.err = &resultErr
+	}
+}
+
+func (client *Client) do(ctx context.Context, method, urlpath string, query url.Values, headers map[string]string, body []byte) ([]*response, error) {
```
Comment (zyga) [inline usersession/client/client.go]: Nitpick, I would document this method and give it a slightly different name to differentiate it from `client.Client.go`, perhaps `doMany`? I will leave it up to you to decide.

File: usersession/client/client.go [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+		return nil, err
+	}
+	for _, resp := range responses {
+		if resp.err == nil {
+			continue
+		}
+		if agentErr, ok := resp.err.(*Error); ok && agentErr.Kind == "service-control" {
+			if errors, ok := agentErr.Value.(map[string]interface{}); ok {
+				for service, failure := range errors {
+					failures = append(failures, ServiceFailure{resp.uid, service, failure.(string)})
```
Comment (zyga) [inline usersession/client/client.go]: I don't understand how we know that the failure is a string here.

File: usersession/client/client_test.go [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    "value": {
+      "service2.service": "failed to stop"
+    }
+  }
+}`))
+	})
+	failures, err := s.cli.ServicesStop(context.Background(), []string{"service1.service", "service2.service"})
+	c.Assert(err, IsNil)
+	c.Check(failures, HasLen, 2)
+	failure1 := failures[0]
```
Comment (zyga) [inline usersession/client/client_test.go]: Mild preference for `failure0`, `failure1` to match array indices.

File: usersession/client/client.go [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+		}
+		return fmt.Errorf("cannot decode %q: %s", buf, err)
+	}
+	return nil
+}
+
+type SessionInfo struct {
+	Version string `json:"version"`
+}
+
```
Comment (zyga) [inline usersession/client/client.go]: Thinking about this request for a second. I think we're missing a RestartSelf API so that client daemon can restart itself after new snapd update. This is not strictly related to this PR but I think we need to address this in general before we start using the daemon -> per user daemon APIs more.

File: usersession/client/client.go [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+func New() *Client {
+	transport := &http.Transport{Dial: dialSessionAgent, DisableKeepAlives: true}
+	return &Client{
+		doer: &http.Client{Transport: transport},
+	}
+}
+
+type Error struct {
```
[the remark this answers is not in the document]
Comment (chipaca) [reply usersession/client/client.go]: at a later date, yes

File: usersession/client/client.go [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+		return nil, err
+	}
+	var (
+		wg        sync.WaitGroup
+		mu        sync.Mutex
+		responses []*response
+	)
+	for _, socket := range sockets {
+		wg.Add(1)
+		go func(socket string) {
```
[the remark this answers is not in the document]
Comment (chipaca) [reply usersession/client/client.go]: in this context (heh) we are receiving a context and using that to make the goroutine cancellable, so I don't think we additionally need a tomb

File: usersession/client/client.go [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+			}
+			if err == nil {
+				defer httpResp.Body.Close()
+				response.statusCode = httpResp.StatusCode
+				response.err = decodeInto(httpResp.Body, &response)
+				response.checkError()
+			}
+			mu.Lock()
+			defer mu.Unlock()
+			responses = append(responses, &response)
```
Comment (chipaca) [inline usersession/client/client.go]: I don't have a strong feeling about this, but I feel it would be slightly more idiomatic to use a channel instead of a waitgroup, a lock, and a list

File: usersession/client/client.go (hunk 4 above)
Comment (chipaca) [reply usersession/client/client.go]: It would be a nice to have, but note the session agent goes away on its own after 30s

File: usersession/client/client.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (jhenstridge) [reply usersession/client/client.go]: Also of note: (1) the meat of the cancellation support is in the HTTP library, which doesn't know about tombs, and (2) the goroutines are all scoped to function calls accepting the context with the wait group ensuring they complete when the function returns.

I had a go at converting it over to a tomb hanging off the passed in context, but it seemed noticeably more verbose to close over the `socket` argument and return something that could be passed to `Tomb.Go`.

File: usersession/client/client.go (hunk 7 above)
Comment (jhenstridge) [reply usersession/client/client.go]: At the moment, there are a few error cases where the goroutine doesn't currently append to the slice: if there are garbage directories under `/run/user`, or if the http library can't create a request instance for some reason.

If we were using tomb, I suppose its `Dead` channel would be a way to know there are no more responses to read (since none of the goroutines are in a position to close the response channel), but that has its own cognitive overhead.

File: usersession/client/client.go [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+		responses []*response
+	)
+	for _, socket := range sockets {
+		wg.Add(1)
+		go func(socket string) {
+			defer wg.Done()
+			uidStr := filepath.Base(filepath.Dir(socket))
+			uid, err := strconv.Atoi(uidStr)
+			if err != nil {
+				logger.Noticef("Socket %q does not appear to be in a valid XDG_RUNTIME_DIR", socket)
```
Comment (pedronis) [inline usersession/client/client.go]: I don't think we want logging at this level which means doMany itself should return successes and errors, which then means that Chipaca suggestion might make sense again

Comment (pedronis) [reply usersession/client/client.go]: a solution could also be a callback approach to what to do about this kind of errors

File: usersession/client/client.go [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+}
+
+type Error struct {
+	Kind    string      `json:"kind"`
+	Value   interface{} `json:"value"`
+	Message string      `json:"message"`
+}
+
+func (e *Error) Error() string {
+	return e.Message
```
Comment (pedronis) [inline usersession/client/client.go]: this bit is not tested

File: usersession/client/client.go [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+		}(socket)
+	}
+	wg.Wait()
+	return responses, nil
+}
+
+func decodeInto(reader io.Reader, v interface{}) error {
+	dec := json.NewDecoder(reader)
+	if err := dec.Decode(v); err != nil {
+		r := dec.Buffered()
```
Comment (pedronis) [inline usersession/client/client.go]: this block is not tested

File: usersession/client/client.go [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+		if resp.err != nil {
+			if err == nil {
+				err = resp.err
+			}
+			continue
+		}
+		var si SessionInfo
+		if decodeErr := json.Unmarshal(resp.Result, &si); decodeErr != nil {
+			if err == nil {
+				err = decodeErr
```
Comment (pedronis) [inline usersession/client/client.go]: this path here is not tested

File: dirs/dirs.go [hunk 12]
```diff
@@ -310,7 +310,7 @@ func SetRootDir(rootdir string) {
 	}
 
 	XdgRuntimeDirBase = filepath.Join(rootdir, "/run/user")
-	XdgRuntimeDirGlob = filepath.Join(rootdir, XdgRuntimeDirBase, "*/")
+	XdgRuntimeDirGlob = filepath.Join(XdgRuntimeDirBase, "*/")
```
[the remark this answers is not in the document]
Comment (pedronis) [reply dirs/dirs.go]: we probably want to split this out in a small PR

File: usersession/client/client.go [hunk 13]
```diff
@@ -0,0 +1,264 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2015-2018 Canonical Ltd
```
Comment (pedronis) [inline usersession/client/client.go]: -2019 ?

File: usersession/client/client_test.go [hunk 14]
```diff
@@ -0,0 +1,257 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2015-2018 Canonical Ltd
```
Comment (pedronis) [inline usersession/client/client_test.go]: 2020 ?

File: usersession/client/client.go (hunk 8 above)
Comment (pedronis) [reply usersession/client/client.go]: to be clear my worry about this logging is also that it might get too noisy and pointless at the same time in some situations

File: dirs/dirs.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (pedronis) [reply dirs/dirs.go]: done in #8070

File: usersession/client/client.go (hunk 8 above)
Comment (jhenstridge) [reply usersession/client/client.go]: I do wonder if it would be better to just silently ignore this case without logging.  This case essentially means that the system administrator has created a non-integer named directory under `/run/user`, and populated it with a file matching the socket path.  It's something that should never be reached, but I agree that it could be noisy the log file if it triggers.

One alternative would be to append another response struct with a dummy value for the UID field (maybe -1?).  But I'm not really sure what the caller could realistically do with that information.

Comment (pedronis) [reply usersession/client/client.go]: yes, after making my initial comment I think this one can be just ignored is a bit too unlikely and the cause would be something local that should be somehow under control anyway

File: usersession/client/client.go [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+			}
+			u := url.URL{
+				Scheme:   "http",
+				Host:     uidStr,
+				Path:     urlpath,
+				RawQuery: query.Encode(),
+			}
+			req, err := http.NewRequest(method, u.String(), bytes.NewBuffer(body))
+			if err != nil {
+				logger.Noticef("Failed to create HTTP request: %v", err)
```
Comment (pedronis) [inline usersession/client/client.go]: this shouldn't happen but should probably surface as an error

File: usersession/client/client.go (hunk 2 above)
Comment (pedronis) [reply usersession/client/client.go]: yes, that seems fragile, it will panic if it isn't

Comment (pedronis) [reply usersession/client/client.go]: same for the other type assertions here

**Comment by pedronis** [comment]
@jhenstridge left a couple more comments,  otherwise the main missing thing is testing what is the behavior if the context used has a deadline and some of the requests but not all of multiple take longer than the deadline

File: usersession/client/client.go (hunk 1 above)
Comment (pedronis) [reply usersession/client/client.go]: yes, something like doMany seems appropriate given the []*response result

File: usersession/client/client.go (hunk 15 above)
Comment (jhenstridge) [reply usersession/client/client.go]: I've rearranged things a bit so this will present as an response error for the uid.  It's not clear how to trigger this code path though.

If we were using Go 1.13's NewRequestWithContext, a nil context would do the trick, but the 1.10 compatible NewRequest + WithContext is a panic rather than an error.

File: usersession/client/client.go (hunk 1 above)
Comment (jhenstridge) [reply usersession/client/client.go]: It is doMany in the current version of the branch.  I forgot to mark this resolved.

File: usersession/client/client.go (hunk 2 above)
Comment (jhenstridge) [reply usersession/client/client.go]: I had updated this code to check the type assertions.  I've added a test to exercise the code with various types of garbage data in the error value.

**Comment by jhenstridge** [comment]
I think I've got a reasonable test for cancellation now.  In the tests, the handlerfunc can use `Request.Host` to distinguish which agent it is mocking.  I have one agent block for 5 seconds or until the client disconnects.  The client reports a result from the well behaved agent, and the error from the malfunctioning agent.

File: usersession/client/client.go [hunk 16]
```diff
@@ -0,0 +1,268 @@
...
+	return info, err
+}
+
+type ServiceFailure struct {
+	Uid     int
+	Service string
+	Error   string
+}
+
+func decodeServiceErrors(uid int, errorValue map[string]interface{}, kind string) []ServiceFailure {
```
Comment (pedronis) [inline usersession/client/client.go]: this should return an error if needed

File: usersession/client/client.go [hunk 17]
```diff
@@ -0,0 +1,268 @@
...
+	errors, ok := errorValue[kind].(map[string]interface{})
+	if !ok {
+		return nil
```
Comment (pedronis) [inline usersession/client/client.go]: this should be some kind of error, "unexpected service-control error value"

File: usersession/client/client.go [hunk 18]
```diff
@@ -0,0 +1,268 @@
...
+	var failures []ServiceFailure
+	for service, reason := range errors {
+		if reasonString, ok := reason.(string); ok {
+			failures = append(failures, ServiceFailure{
+				Uid:     uid,
+				Service: service,
+				Error:   reasonString,
+			})
+		} else {
+			logger.Noticef("Could not decode %s failure for %q: expected string, but got %T", kind, service, reason)
```
Comment (pedronis) [inline usersession/client/client.go]: same

File: usersession/client/client.go [hunk 19]
```diff
@@ -0,0 +1,268 @@
...
+		return nil, nil, err
+	}
+	for _, resp := range responses {
+		if resp.err == nil {
+			continue
+		}
+		if agentErr, ok := resp.err.(*Error); ok && agentErr.Kind == "service-control" {
+			if errorValue, ok := agentErr.Value.(map[string]interface{}); ok {
+				startFailures = append(startFailures, decodeServiceErrors(resp.uid, errorValue, "start-errors")...)
+				stopFailures = append(stopFailures, decodeServiceErrors(resp.uid, errorValue, "stop-errors")...)
```
Comment (pedronis) [inline usersession/client/client.go]: then if the decodeServiceErrors return an error this code should just end up using the path below with using resp.err as is

File: usersession/client/client.go [hunk 20]
```diff
@@ -0,0 +1,268 @@
...
+
+			u := url.URL{
+				Scheme:   "http",
+				Host:     uidStr,
+				Path:     urlpath,
+				RawQuery: query.Encode(),
+			}
+			req, err := http.NewRequest(method, u.String(), bytes.NewBuffer(body))
+			if err != nil {
+				response.err = err
```
Comment (pedronis) [inline usersession/client/client.go]: maybe wrap this in  fmt.Errorf("internal error: %v", err)

File: usersession/client/client.go (hunk 16 above)
Comment (jhenstridge) [reply usersession/client/client.go]: I've reorganised this function to add an error return.  Note that in this code path we've also got an error message from the agent in the body.  At present, the Services*() method return a single error, so I thought it would be better to return that error to the caller.

I guess the alternative would be to switch over to reporting a list of errors rather than the first, but that kind of thing doesn't seem to be common elsewhere in the code base.
